### ISSUE TITLE
no-invalid-regexp should catch non-cross-platform regex features

### ISSUE THREAD

**Issue by Daniel15** [eslint/eslint#11931]
**What rule do you want to change?**
`no-invalid-regexp`

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
const namedCaptureGroupsExample = /^(?<foo>.+)/;
const lookBehindExample = /(?<=[a-z])(?=[A-Z])/g;
```

This uses regex feature (named capture groups and lookbehind) that are only supported in some browsers: https://www.smashingmagazine.com/2019/02/regexp-features-regular-expressions/. Firefox actually throws a syntax error for this.

**What does the rule currently do for this code?**
Nothing

**What will the rule do after it's changed?**
Error: Named capture groups are not supported in all browsers
Error: Lookbehind is not supported in all browsers

**Are you willing to submit a pull request to implement this change?**
Maybe, if this proposal sounds OK

**Comment by platinumazure** [comment]
Hi @Daniel15, thanks for the issue!

I don't think we can put in options for specific runtimes, but we might be able to put in options for allowing or disallowing specific regex features. Do you have any thoughts on what options we should add?

**Comment by Daniel15** [comment]
> Do you have any thoughts on what options we should add?

At least all the features that Firefox doesn't currently support. I know of the two examples in my post, but there could be others.

At FB, we've had breakages in some internal tools due to usage of regex features that Firefox doesn't support. Unfortunately, these are syntax errors in Firefox, meaning they break the entire script at parse time 😢

**Comment by mysticatea** [comment]
You may be interested in [eslint-plugin-es](https://mysticatea.github.io/eslint-plugin-es/). It allows us to disable each syntax individually.

**Comment by Daniel15** [comment]
Thank you @mysticatea - That is perfect!! Just enabled these rules in my ESLint config:
```
'es/no-regexp-lookbehind-assertions'
'es/no-regexp-named-capture-groups'
'es/no-regexp-s-flag'
'es/no-regexp-unicode-property-escapes'
```
